### PR TITLE
Formalize usage stats for analytics

### PR DESCRIPTION
**Pull request by nik9000**
This moves the usage statistics gathering from the `AnalyticsPlugin`
into an `AnalyicsUsage`, removing the static state. It also checks the
license level when parsing all analytics aggregations. This is how we
were checking them before but we did it in an easy to forget way. This
way is slightly simpler, I think.

### CODE REVIEW

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsUsage.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Tracks usage of the Analytics aggregations.
+ */
+public class AnalyticsUsage {
+    private final AtomicLong boxplotUsage = new AtomicLong(0);
+    private final AtomicLong cumulativeCardUsage = new AtomicLong(0);
+    private final AtomicLong stringStatsUsage = new AtomicLong(0);
+    private final AtomicLong topMetricsUsage = new AtomicLong(0);
+
+    public <C, T> ContextParser<C, T> trackBoxplot(ContextParser<C, T> realParser) {
```
Comment (not-napoleon) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsUsage.java]: Having a `trackFoo` and `getFooUsage` for each aggregation seems like a lot of boiler plate to me.  If I was writing this, I'd make an enum of things to track, have an `EnumMap` of `AtomicLong`s, and just one `track` and one `getUsage` method which accept that enum as an additional argument.  `EnumMap` uses an efficient array implementation, so we wouldn't be paying much in overhead for that.   This would mean that adding a new counter to this class would just. require adding an entry to the enum, rather than also copying and modifying the track and get methods.  Thoughts?

**Review by not-napoleon — CHANGES_REQUESTED** [review]
Looks good overall, and much cleaner than what we had before.  I left one suggestion I'd like to either implement or know why we're not.  Thanks!

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsUsage.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/AnalyticsUsage.java]: I like it!

**Comment by nik9000** [comment]
@not-napoleon I updated the usage tracker to to be the way you described. I *didn't* change the response to use an `EnumMap` to longs because I feel like the explicit ceremony there helps to make sure that you don't make mistakes working with the wire protocol.

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregationSortSource.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+package org.elasticsearch.search.aggregations;
+
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.util.Comparator;
+
+/**
+ * Compare two buckets by their ordinal.
+ */
+public interface AggregationSortSource<A extends Aggregator, R extends Aggregation> {
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/AggregationSortSource.java]: Did you mean to include this in this PR? it seems unrelated.

**Review by not-napoleon — APPROVED** [review]
LGTM, thanks for the revision.

File: server/src/main/java/org/elasticsearch/search/aggregations/AggregationSortSource.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/AggregationSortSource.java]: > Did you mean to include this in this PR? it seems unrelated.

Nope. Sneaky `git add .`! I'll remove.

**Comment by nik9000** [comment]
Thanks @not-napoleon!
